unittesting/core/st3/fast_runner.py

```diff
@@ -0,0 +1,167 @@
+        startTime = time.time()
+
+        def _start_testing():
+            with warnings.catch_warnings():
+                if self.warnings:
+                    # if self.warnings is set, use it to filter all the warnings
+                    warnings.simplefilter(self.warnings)
+                    # if the filter is 'default' or 'always', special-case the
+                    # warnings from the deprecated unittest methods to show them
+                    # no more than once per module, because they can be fairly
+                    # noisy.  The -Wd and -Wa flags can be used to bypass this
+                    # only when self.warnings is None.
+                    if self.warnings in ['default', 'always']:
+                        warnings.filterwarnings(
+                            'module',
+                            category=DeprecationWarning,
+                            message='Please use assert\\w+ instead.')
+                startTestRun = getattr(result, 'startTestRun', None)
+                if startTestRun is not None:
+                    startTestRun()
+                try:
+                    deferred = test(result)
+                    _continue_testing(deferred)
+
+                except Exception as e:
+                    _handle_error(e)
+
+        def _continue_testing(deferred, send_value=None, throw_value=None):
+            try:
+                if throw_value:
+                    condition = deferred.throw(throw_value)
+                else:
+                    condition = deferred.send(send_value)
+
+                if callable(condition):
+                    defer(0, _wait_condition, deferred, condition)
+                elif isinstance(condition, dict) and "condition" in condition and \
+                        callable(condition["condition"]):
+                    period = condition.get("period", DEFAULT_CONDITION_POLL_TIME)
+                    defer(period, _wait_condition, deferred, **condition)
+                elif isinstance(condition, int):
+                    defer(condition, _continue_testing, deferred)
+                elif condition == AWAIT_WORKER:
+                    run_on_worker(
+                        partial(defer, 0, _continue_testing, deferred)
+                    )
+                else:
+                    defer(0, _continue_testing, deferred)
+
+            except StopIteration:
+                _stop_testing()
+                self.finished = True
+
+            except Exception as e:
+                _handle_error(e)
+
+        def _wait_condition(
+            deferred, condition,
+            period=DEFAULT_CONDITION_POLL_TIME,
+            timeout=DEFAULT_CONDITION_TIMEOUT,
+            start_time=None
+        ):
+            if start_time is None:
+                start_time = time.time()
+
+            try:
+                send_value = condition()
+            except Exception as e:
+                _continue_testing(deferred, throw_value=e)
```

Comment:
Please help me to understand. Why we cannot do `defer(0, _continue_testing, deferred, throw_value=e)`?

Reply:
Oh, was it because of the coverage issue?

Reply:
Okay, you're trying to modify the behavior to make the generalization/abstraction easier. That's cheating.

Now, sure both runners *look* similar but the execution flow is completely different. For sync programs, we can look at conditional branching (e.g. ifs) or method delegation to get the 'flow'. But this is async code, only matters *where* we run and *when*. The code we're looking at is an 'executor' for a (test) program.

Here I want `fn()` behavior which is sync and blocking, `defer(0, fn)` is not sync but cooperative (in the sense that other tasks in the queue can run before fn) but still blocking (in the sense that all these tasks marked as to-be-run-immediate run in a chain without gaps). `defer(10, fn)` is not sync and not blocking.

Please look at the `DeferrableTestSuite` in `suite.py`. We should remove all plain `yield`s in `run()`. What is your preferred way of writing this?

Reply:
It might come down to our different expectations about synchronization of the test cases.	
The major difference between the original runner and the new runner is whether we run the next test function/ test case in a deferred manner. It is also related to the plain `yield`s in `DeferrableTestSuite`.

```diff
@@ -48,7 +50,7 @@ class DeferringTextTestRunner(TextTestRunner):
                     startTestRun()
                 try:
                     deferred = test(result)
-                    defer(10, _continue_testing, deferred)
+                    _continue_testing(deferred)
```

When I first wrote it, I thought it would be just easier to defer everywhere. I have to admit that it may be not very efficient and strictly necessary. The idea here is to make sure the UI has time to respond and the tests won't block the UI. The name of the runner also suggests that it runs tests in a deferring manner. This is not guaranteed that the code will be synchronous.

I understand your point of keeping the code clean and simple. But if we look at the diff of the original runner and the fast runner, there are actually not many lines which are different. It is hard to imagine why we should not refactor the runners to make it more customizable. The real issue here is whether we should defer between tests. The setting `fast_timings` actually does not explain explicitly what it is being faster. Perhaps we need a more specific setting to control whether the coroutine is executed asynchronously and by how much. I am still a bit leaning towards `defer_timeout` with the understanding that `defer_timeout = 0` means `fn()` rather than `timeout(fn, 0)` in some of the lines.

The actual operation is actually hidden from the users. For the users, that number just means the number of seconds to wait. `0` just means "do not wait".

If you are worried about breakage, you have already protected yourself by writing the corresponding tests.